### PR TITLE
fix: local OWN_CANISTER_ID

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
The nns-dapp configuration is wrong for local deployments:  The local nns-dapp canister ID is overridden.

# Changes
- Use the term `OWN_CANISTER_ID` consistently to make this easier to debug.  Previously sometimes `CANISTER_ID` was used and that gives many false hits when searching.
- Make the canister IDs provided by dfx override those in configuration files.

# Tests
- I added a test for teh output of config.sh for local, with two different deployments of the nns-dapp.
  - Note: There is an existing test that covers mainnet and the app subnet, but these both have fixed canister IDs.  The test needed to be adapted a little to cope with a variable canister ID.
- I read through all uses of these variables I could find and am reasonably confident on that basis that these changes are correct, not just that the tests pass.  Additional vigilance will also be needed in the next release to make sure that the arguments are correct there.

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

File: config.sh [hunk 1]
```diff
@@ -160,11 +152,11 @@ local_deployment_data="$(
 : "- construct ledger and governance canister URLs"
 # TODO: I believe that the following can be discarded now.
 json=$(HOST=$(dfx-canister-url --network "$DFX_NETWORK" --type api) jq -s --sort-keys '
-  (.[0].defaults.network.config // {}) * .[1] * (.[2].config // {}) |
+  (.[0].defaults.network.config // {}) * (.[1].config // {}) * .[2] |
   .DFX_NETWORK = env.DFX_NETWORK |
   . as $config |
   .HOST=env.HOST
-    ' dfx.json <(echo "$local_deployment_data") <(network_config))
+    ' dfx.json <(network_config) <(echo "$local_deployment_data"))
```
Comment (bitdivine) [inline config.sh:159]: This is the actual fix.

File: config.sh [hunk 2]
```diff
@@ -48,9 +48,9 @@ local_deployment_data="$(
   : "- may be set by dfx as an env var"
   : "- may be deployed locally"
   LOCALLY_DEPLOYED_NNS_CANISTER_ID="$(dfx canister --network "$DFX_NETWORK" id nns-dapp 2>/dev/null || true)"
-  test -n "${CANISTER_ID:-}" || CANISTER_ID="$LOCALLY_DEPLOYED_NNS_CANISTER_ID"
-  export CANISTER_ID
-  test -n "${CANISTER_ID:-}" || unset CANISTER_ID
+  test -n "${OWN_CANISTER_ID:-}" || OWN_CANISTER_ID="$LOCALLY_DEPLOYED_NNS_CANISTER_ID"
+  export OWN_CANISTER_ID
+  test -n "${OWN_CANISTER_ID:-}" || unset OWN_CANISTER_ID
```
Comment (bitdivine) [inline config.sh:53]: This `CANISTER_ID` is hard to search for in logs as it appears in other places and as a substring of other variables.  `OWN_CANISTER_ID` is unique.  We might even switch to `CANISTER_ID_NNS_DAPP` as used by `dfx`.  https://internetcomputer.org/docs/current/references/cli-reference/dfx-envars  But that would touch more files and so belongs in another PR.

File: dfx.json [hunk 3]
```diff
@@ -299,7 +299,6 @@
     },
     "local": {
       "config": {
-        "FETCH_ROOT_KEY": true,
         "HOST": "http://localhost:8080",
```
Comment (bitdivine) [inline dfx.json:302]: Not needed as it matches the default and increases noise.

File: config.test [hunk 4]
```diff
@@ -16,7 +16,17 @@ clean_config() {
...
+  echo "Compare generated config with local deployments"
+  # Note: Unlike on mainnet, canister IDs can change.  The test will adapt the expected values with the actual canister ID.
+  # We will deploy twice to make sure that the local canister ID changes.
+  dfx start --clean --background
+  for ((i = 0; i < 2; i++)); do
+    dfx canister create nns-dapp
+    scripts/nns-dapp/test-config --network local
+    dfx canister delete nns-dapp --yes
+  done
+)
```
Comment (bitdivine) [inline config.test:29]: We test local deployments twice to make sure that we test different canister IDs.

File: config.sh (hunk 2 above)
Comment (lmuntaner) [reply config.sh:53]: I also like the idea of `CANISTER_ID_NNS_DAPP`. It's clearer.

File: config.sh [hunk 5]
```diff
@@ -120,16 +120,9 @@ local_deployment_data="$(
   STATIC_HOST="$(dfx-canister-url --network "$DFX_NETWORK" --type static)"
   export API_HOST STATIC_HOST
 
-  : "Determine whether we need to fetch the root key"
```
Comment (lmuntaner) [inline config.sh:123]: Why did you remove this?

File: config.sh (hunk 1 above)
Comment (lmuntaner) [reply config.sh:159]: Thanks for pointing it out!

**Review by lmuntaner — APPROVED** [review]
LGTM! thanks!

File: config.sh (hunk 5 above)
Comment (bitdivine) [reply config.sh:123]: It gives the wrong answer for e.g. the app subnet "network".  `scripts/network-config` gives the right answer.

Comment (bitdivine) [reply config.sh:123]: Resolving to merge but happy to discuss further if wanted.
